### DISCUSSION TITLE
Use Project ID for the type library ID?

### DISCUSSION THREAD

**Discussion by GCuser99** [twinbasic/twinbasic#1498] [Q&A]
Hi, I have been testing twinBasic ActiveX DLL's for use in MS Office apps and feel like I have been making much progress... but not being an expert, I'm wondering what to do with the following project option:

![image](https://user-images.githubusercontent.com/26237126/219512640-c5572218-dec2-4e3b-a0bb-3dc8f4f2858e.png)

Under what usage case(s) should I set that option to "yes"?

**Comment by bclothier** [comment]
TBH, I always set it to `Yes` because having a stable type library ID helps a lot in debugging. For the same reasons, I'm usually explicit with my GUIDs everywhere. I know in VB6 they hide all that GUID magic but that makes me nervous about polluting the registry with random & semi-duplicated values, so I'd rather have same GUIDs explicitly defined and persisted to avoid those problems.

**Reply by GCuser99** [reply]
Yeah, I notice in your [TwinBasicSevenZip](https://github.com/bclothier/TwinBasicSevenZip) project you had it set to yes. 

Under what scenario would you leave it to default to "no"?

**Reply by bclothier** [reply]
To be honest, I'm not aware of a situation why it should be "no". I'm distrustful of anything that's trying to do things for me behind the scene (ironic I know but that's because I've been bitten by situations where a cryptic error became apparent only after you know what's going on behind the scene --- see Joel Spolsky's Law of Leaky Abstraction).

**Reply by GCuser99** [reply]
Thanks @bclothier - I'll set it to "yes", in the absence of a reason to set it to "no".  :-)

**Comment by GCuser99** [comment]
But as a point of clarification, when the project option box is left unchecked, with "inherited from [Defaults]" stipulated in the upper right corner, what is the default? Is it the greyed-out "No" in the dropdown? Where do I find what's in "[Defaults]"?

**Comment by bclothier** [comment]
Yes, that is true for all options - if you uncheck the option box on the left, it will display the default value. You can see that by checking the option value, then changing the value to something else, then uncheck it - it will change back to the default value.

I _think_ that was inherited from the original VS Code design where a defaults section was a part of the code workspace but that is no longer exposed.

**Comment by wqweto** [comment]
@WaynePhillipsEA Can we use this option (or separate one) to get some deterministic IID/CLSID generation scheme in TB compiler?

The idea is for unspecified GUIDs on interfaces/classes which get into the public typelib of the Ax DLL/Control to generate the GUID from hash of LIBID + objtype + objname

The generation can use SHA-256 (or even MD-5) like `LEFT(HASH("092ceb4e-aa99-4e0d-b88f-75b6ff4253e4" + "interace" + "ISomething"), 16)` then "fix" this byte sequence to be a valid v4 UUID (or similar) and use the result when explicit value for the interface/class is missing.

This will stabilize x86 vs x64 typelibs too to prevent nasty IID/CLSID mismatches when both are registered.

**Reply by bclothier** [reply]
For the record, I'd prefer the solution resulting in explicitly generating the GUID so that it's right there in front of your eyes, no more magic behind the scene.

**Reply by wqweto** [reply]
My suggestion is about the case when one does not specify explicit GUIDs so that the random generated GUIDs are stable i.e. do not change between recompilations and between x86 to x64 targets (the same way explicit GUIDs do not change already) because there is an issue if a same LIBID typelib is registered both 32-bit and 64-bit but with different IIDs/CLSIDs.

The point is that if LIBID is explicit then regenerating non-explicit GUIDs non-deterministically is a buggy behavior so a deterministric strategy for non-explicit GUID is non-optional for the compiler IMO.

**Reply by WaynePhillipsEA** [reply]
Perhaps a compromise: if the IIDs are not defined explicitly, generate deterministic ones, which should be visible via a CodeLens line in the IDE in the place where the explicit ones would normally be found (i.e. above the Class definition)

**Reply by bclothier** [reply]
I understand but what I'm suggesting is that when creating a new class (or possibly at the build time, the compiler detects missing GUID and adds it), that it will do so by adding the attributes at the top of the class so that it's visible. Not showing it (even if it's somehow preserved between compiles) seems to me to be a recipe to confusion -- the law of leaky abstraction will bite the user.

The CodeLens that can be clicked to become an attribute is certainly an approach worth considering. It just can't be out of sight, out of mind.

**Reply by bclothier** [reply]
One caveat I thought of after the above, is that if CodeLen idea means having a text document that's not the same as shown in code editor (as is currently the case w/ VBx's documents), then I might need to push back on that idea. From the POV of maintainability having a document that's different between the code pane and saved on disk is very bad for source code control. We should avoid having any difference in text that's visible via the code pane if possible, so adding explicit attributes would be preferable to some magical storage. Those should be easily folded away if they don't care about details but the text must be the same between the code pane and the disk.

**Reply by wqweto** [reply]
Aha, so Add Class wizard to generate the GUID -- possible. Yes, tooling is one way to attack this problem but both solutions can coexist for the case when code is imported without attributes.

Another option is to refuse compilation when GUIDs are not explicitly generated for public classes/interfaces which is too drastic and is already considered as a project option.

For the case when the binary **has** to be produced by the compiler and in the process there are GUIDs to be randomly generated, then using a deterministric approach is my suggestion for the sheer "randomness" of this operation.

Edit: This can be part of the "reporoducable builds" feature too, if TB will aim at such thing.

Edit2: Somehow missed Wayne's suggestion above which would be useful indeed.
